### PR TITLE
feat: pagination for survey list

### LINKED ISSUE #79: [P0] Survey list pagination API
This should be in descending order with paginated by the size of 8

### PR DESCRIPTION
**Pull request by bezzang-dev**
**Added**
- `com/thesurvey/api/dto/response/SurveyListPageDto.java` to response `getAllSurvey` in `SurveyController`
- `com/thesurvey/api/dto/response/SurveyPageDto.java` to save paging survey info
- `com/thesurvey/api/dto/request/PageableSwagger.java`to custom "모든 설문조사 조회" page in swagger
-  test code to pagination in `SurveyControllerTest` and `SurveyServiceTest`
- Implemented validate invalid request to pagination API in `SurveyService`

**Modified**
- The paging has been applied to the `getAllSurvey` method in `SurveyController`

Clients can only use the `page` param to request and page `size` is fixed at **8**.
Clients can request the server as follows: 
`/v1/surveys?page=1`

<details> 
<summary> response pagination API example </summary>
<pre> 
"surveys": [
        {
            "surveyId": "ede63ee6-c63c-43da-9173-6a1e269a17a3",
            "authorId": 1,
            "title": "Sample Survey10",
            "description": "A simple survey",
            "startedDate": "2023-04-22T14:25:00",
            "endedDate": "2023-04-27T23:59:59",
            "modifiedDate": "2023-04-22T01:43:57.498783",
            "certificationTypes": [
                "KAKAO",
                "NAVER"
            ]
        }
    ],
    "page": 1,
    "totalSurveys": 1,
    "totalPages": 1
}
</pre>
</details>

This PR closes #79

### CODE REVIEW

**Comment by seiwonpark** [comment]
_(Did I implement pagination or paging..??, Which term is correct??)_ -> Pagination is correct See [this](https://developers.google.com/search/docs/specialty/ecommerce/pagination-and-incremental-page-loading?hl=ko)

**Comment by seiwonpark** [comment]
```
 
{
    "content": [ // The data included in the current page
        {
            "surveyId": "cd21679c-bb69-46e4-8453-528516452d0c",
            "authorId": 1,
            "title": "Sample Survey",
            "description": "A simple survey",
            "startedDate": "2023-04-21T14:25:00",
            "endedDate": "2023-04-27T23:59:59",
            "modifiedDate": "2023-04-21T04:03:45.163133",
            "certificationTypes": [
                "KAKAO",
                "NAVER"
            ]
        }
    ],
    "pageable": { // Information about the page
        "sort": {
            "empty": true,
            "unsorted": true,
            "sorted": false
        },
        "offset": 0, // indicates the starting position of the current page
        "pageNumber": 0,
        "pageSize": 8,
        "paged": true,
        "unpaged": false
    },
    "last": true, // Indicates whether the current page is the last page.
    "totalElements": 1, // The total number of pages.
    "totalPages": 1, // The total number of pages.
    "number": 0, // The current page number.
    "sort": {
        "empty": true,
        "unsorted": true,
        "sorted": false
    },
    "first": true, // Indicates whether the current page is the first page.
    "size": 8, // The page size.
    "numberOfElements": 1,
    "empty": false
}
```

I see too much duplicated values. Could you cut off some? IDK why values like `"first": true,` is needed.

File: api/src/main/java/com/thesurvey/api/controller/SurveyController.java [hunk 1]
```diff
@@ -51,8 +54,8 @@ public SurveyController(SurveyService surveyService,
         @ApiResponse(code = 404, message = "요청한 리소스 찾을 수 없음")
     })
     @GetMapping
-    public ResponseEntity<List<SurveyResponseDto>> getAllSurvey() {
-        return ResponseEntity.ok(surveyService.getAllSurvey());
+    public ResponseEntity<Page<SurveyListPageDto>> getAllSurvey(@PageableDefault(page = 0, size = 8) Pageable pageable) {
```
Comment (seiwonpark) [inline api/src/main/java/com/thesurvey/api/controller/SurveyController.java]: **!!!**

array

Comment (seiwonpark) [reply api/src/main/java/com/thesurvey/api/controller/SurveyController.java]: How about setting entry page to 1?

File: api/src/main/java/com/thesurvey/api/dto/request/CustomPageable.java [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+
+import io.swagger.annotations.ApiModel;
+import io.swagger.annotations.ApiModelProperty;
+import lombok.Getter;
+
+@Getter
+@ApiModel
+public class CustomPageable {
+
+    @ApiModelProperty(name = "페이지 수", example = "가져올 페이지의 번호를 나타냅니다.", notes = "0 부터 시작하며, 유효하지 않은 값은 0으로 설정됩니다.")
```
Comment (seiwonpark) [inline api/src/main/java/com/thesurvey/api/dto/request/CustomPageable.java]: `example` should be the actual value. This is not description

**Comment by bezzang-dev** [comment]
I will add a response DTO to modify the API's response and include only the necessary information.

File: api/src/main/java/com/thesurvey/api/exception/ErrorMessage.java [hunk 3]
```diff
@@ -24,7 +24,10 @@ public enum ErrorMessage {
     NO_ONLY_WHITESPACE("공백만 입력할 수 없습니다."),
     CREATOR_CANNOT_ANSWER("설문조사 작성자는 자신의 설문조사에 응답할 수 없습니다."),
     NOT_SURVEY_QUESTION("해당 설문조사의 질문이 아닙니다."),
-    INVALID_REQUEST("유효하지 않은 요청입니다");
+    INVALID_REQUEST("유효하지 않은 요청입니다"),
+    INVALID_PAGE_SIZE("페이지 크기는 8 이어야 합니다."),
```
Comment (seiwonpark) [inline api/src/main/java/com/thesurvey/api/exception/ErrorMessage.java]: Why would client need to know `pageSize`? This should be fixed value. `PAGE_NOT_FOUND` should be enough

File: api/src/main/java/com/thesurvey/api/exception/ErrorMessage.java [hunk 4]
```diff
@@ -24,7 +24,10 @@ public enum ErrorMessage {
...
+    INVALID_PAGE_SORT("페이지 정렬은 기본 값이어야 합니다."),
```
Comment (seiwonpark) [inline api/src/main/java/com/thesurvey/api/exception/ErrorMessage.java]: Same above. Client should not be able to set anything for pagination except request parameter `page` which indicates actual page number. Seems like this error message is for some client request that could change the method of sorting policy

File: api/src/test/java/com/thesurvey/api/controller/SurveyControllerTest.java [hunk 5]
```diff
@@ -226,6 +226,37 @@ void testUpdateSurvey() throws Exception {
...
+                .endedDate(LocalDateTime.now(ZoneId.of("Asia/Seoul")).plusDays(2))
+                .certificationTypes(List.of(CertificationType.GOOGLE))
+                .questions(List.of(questionRequestDto))
+                .build();
+            mockCreateSurvey(surveyRequestDto);
+        }
+
+        // when
+        MvcResult result = mockMvc.perform(get("/surveys")
+                .param("page", "1")
```
Comment (seiwonpark) [inline api/src/test/java/com/thesurvey/api/controller/SurveyControllerTest.java]: array

Comment (seiwonpark) [reply api/src/test/java/com/thesurvey/api/controller/SurveyControllerTest.java]: And this starts from `1`!

File: api/src/main/java/com/thesurvey/api/exception/ErrorMessage.java (hunk 4 above)
Comment (seiwonpark) [reply api/src/main/java/com/thesurvey/api/exception/ErrorMessage.java]: Oh okay this is  for **filtering** option. Resolved.

**Review by seiwonpark — CHANGES_REQUESTED** [review]
Please update your code and re-request.

**Review by seiwonpark — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**